# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/bindings/TypedArray.h
# Commit: 6880cf8c51bc
# Full Hash: 6880cf8c51bc665b19b3d14b9fe4edb7960f527e
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2021-08-13 09:27:46
# Regressor Bug: 1720422
# File Overlap Count: 1
# Description:
#   Bug 1720422 - Switch DOM bindings typed arrays to using JSAPI templates r=jonco,smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D119858
# ==============================================================================

diff -r 23cc649c6e14 -r 6880cf8c51bc dom/bindings/TypedArray.h
--- a/dom/bindings/TypedArray.h	Thu Aug 12 20:44:18 2021 +0000
+++ b/dom/bindings/TypedArray.h	Thu Aug 12 20:44:18 2021 +0000
@@ -14,7 +14,7 @@
 #include "js/experimental/TypedData.h"  // js::Unwrap(Ui|I)nt(8|16|32)Array, js::Get(Ui|I)nt(8|16|32)ArrayLengthAndData, js::UnwrapUint8ClampedArray, js::GetUint8ClampedArrayLengthAndData, js::UnwrapFloat(32|64)Array, js::GetFloat(32|64)ArrayLengthAndData, JS_GetArrayBufferViewType
 #include "js/GCAPI.h"                   // JS::AutoCheckCannotGC
 #include "js/RootingAPI.h"              // JS::Rooted
-#include "js/ScalarType.h"              // js::Scalar::Type
+#include "js/ScalarType.h"              // JS::Scalar::Type
 #include "js/SharedArrayBuffer.h"
 #include "mozilla/Attributes.h"
 #include "mozilla/dom/BindingDeclarations.h"
@@ -30,11 +30,10 @@
  * a subclass of the base class that supports creation of a relevant typed array
  * or array buffer object.
  */
-template <typename T, JSObject* UnwrapArray(JSObject*),
-          void GetLengthAndDataAndSharedness(JSObject*, size_t*, bool*, T**)>
+template <class ArrayT>
 struct TypedArray_base : public SpiderMonkeyInterfaceObjectStorage,
                          AllTypedArraysBase {
-  typedef T element_type;
+  using element_type = typename ArrayT::DataType;
 
   TypedArray_base()
       : mData(nullptr), mLength(0), mShared(false), mComputed(false) {}
@@ -49,7 +48,7 @@
   }
 
  private:
-  mutable T* mData;
+  mutable element_type* mData;
   mutable uint32_t mLength;
   mutable bool mShared;
   mutable bool mComputed;
@@ -57,7 +56,7 @@
  public:
   inline bool Init(JSObject* obj) {
     MOZ_ASSERT(!inited());
-    mImplObj = mWrappedObj = UnwrapArray(obj);
+    mImplObj = mWrappedObj = ArrayT::unwrap(obj).asObject();
     return inited();
   }
 
@@ -107,7 +106,7 @@
     return mShared;
   }
 
-  inline T* Data() const {
+  inline element_type* Data() const {
     MOZ_ASSERT(mComputed);
     return mData;
   }
@@ -119,7 +118,7 @@
   // direct pointer to the actual data with no copying. If the provided buffer
   // is not large enough, nullptr will be returned. If bufSize is at least
   // JS_MaxMovableTypedArraySize(), the data is guaranteed to fit.
-  inline T* FixedData(uint8_t* buffer, size_t bufSize) const {
+  inline element_type* FixedData(uint8_t* buffer, size_t bufSize) const {
     MOZ_ASSERT(mComputed);
     return JS_GetArrayBufferViewFixedData(mImplObj, buffer, bufSize);
   }
@@ -133,7 +132,9 @@
     MOZ_ASSERT(inited());
     MOZ_ASSERT(!mComputed);
     size_t length;
-    GetLengthAndDataAndSharedness(mImplObj, &length, &mShared, &mData);
+    JS::AutoCheckCannotGC nogc;
+    mData =
+        ArrayT::fromObject(mImplObj).getLengthAndData(&length, &mShared, nogc);
     MOZ_RELEASE_ASSERT(length <= INT32_MAX,
                        "Bindings must have checked ArrayBuffer{View} length");
     mLength = length;
@@ -154,22 +155,18 @@
   TypedArray_base(const TypedArray_base&) = delete;
 };
 
-template <typename T, JSObject* UnwrapArray(JSObject*),
-          T* GetData(JSObject*, bool* isShared, const JS::AutoRequireNoGC&),
-          void GetLengthAndDataAndSharedness(JSObject*, size_t*, bool*, T**),
-          JSObject* CreateNew(JSContext*, size_t)>
-struct TypedArray
-    : public TypedArray_base<T, UnwrapArray, GetLengthAndDataAndSharedness> {
- private:
-  typedef TypedArray_base<T, UnwrapArray, GetLengthAndDataAndSharedness> Base;
+template <class ArrayT>
+struct TypedArray : public TypedArray_base<ArrayT> {
+  using Base = TypedArray_base<ArrayT>;
+  using element_type = typename Base::element_type;
 
- public:
   TypedArray() = default;
 
   TypedArray(TypedArray&& aOther) = default;
 
   static inline JSObject* Create(JSContext* cx, nsWrapperCache* creator,
-                                 uint32_t length, const T* data = nullptr) {
+                                 uint32_t length,
+                                 const element_type* data = nullptr) {
     JS::Rooted<JSObject*> creatorWrapper(cx);
     Maybe<JSAutoRealm> ar;
     if (creator && (creatorWrapper = creator->GetWrapperPreserveColor())) {
@@ -180,12 +177,12 @@
   }
 
   static inline JSObject* Create(JSContext* cx, uint32_t length,
-                                 const T* data = nullptr) {
+                                 const element_type* data = nullptr) {
     return CreateCommon(cx, length, data);
   }
 
   static inline JSObject* Create(JSContext* cx, nsWrapperCache* creator,
-                                 Span<const T> data) {
+                                 Span<const element_type> data) {
     // Span<> uses size_t as a length, and we use uint32_t instead.
     if (MOZ_UNLIKELY(data.Length() > UINT32_MAX)) {
       JS_ReportOutOfMemory(cx);
@@ -194,7 +191,7 @@
     return Create(cx, creator, data.Length(), data.Elements());
   }
 
-  static inline JSObject* Create(JSContext* cx, Span<const T> data) {
+  static inline JSObject* Create(JSContext* cx, Span<const element_type> data) {
     // Span<> uses size_t as a length, and we use uint32_t instead.
     if (MOZ_UNLIKELY(data.Length() > UINT32_MAX)) {
       JS_ReportOutOfMemory(cx);
@@ -205,47 +202,41 @@
 
  private:
   static inline JSObject* CreateCommon(JSContext* cx, uint32_t length,
-                                       const T* data) {
-    JSObject* obj = CreateNew(cx, length);
-    if (!obj) {
+                                       const element_type* data) {
+    auto array = ArrayT::create(cx, length);
+    if (!array) {
       return nullptr;
     }
     if (data) {
       JS::AutoCheckCannotGC nogc;
       bool isShared;
-      T* buf = static_cast<T*>(GetData(obj, &isShared, nogc));
+      element_type* buf = array.getData(&isShared, nogc);
       // Data will not be shared, until a construction protocol exists
       // for constructing shared data.
       MOZ_ASSERT(!isShared);
-      memcpy(buf, data, length * sizeof(T));
+      memcpy(buf, data, length * sizeof(element_type));
     }
-    return obj;
+    return array.asObject();
   }
 
   TypedArray(const TypedArray&) = delete;
 };
 
-template <JSObject* UnwrapArray(JSObject*),
-          void GetLengthAndDataAndSharedness(JSObject*, size_t*, bool*,
-                                             uint8_t**),
-          js::Scalar::Type GetViewType(JSObject*)>
-struct ArrayBufferView_base
-    : public TypedArray_base<uint8_t, UnwrapArray,
-                             GetLengthAndDataAndSharedness> {
+template <JS::Scalar::Type GetViewType(JSObject*)>
+struct ArrayBufferView_base : public TypedArray_base<JS::ArrayBufferView> {
  private:
-  typedef TypedArray_base<uint8_t, UnwrapArray, GetLengthAndDataAndSharedness>
-      Base;
+  using Base = TypedArray_base<JS::ArrayBufferView>;
 
  public:
-  ArrayBufferView_base() : Base(), mType(js::Scalar::MaxTypedArrayViewType) {}
+  ArrayBufferView_base() : Base(), mType(JS::Scalar::MaxTypedArrayViewType) {}
 
   ArrayBufferView_base(ArrayBufferView_base&& aOther)
       : Base(std::move(aOther)), mType(aOther.mType) {
-    aOther.mType = js::Scalar::MaxTypedArrayViewType;
+    aOther.mType = JS::Scalar::MaxTypedArrayViewType;
   }
 
  private:
-  js::Scalar::Type mType;
+  JS::Scalar::Type mType;
 
  public:
   inline bool Init(JSObject* obj) {
@@ -257,49 +248,23 @@
     return true;
   }
 
-  inline js::Scalar::Type Type() const {
+  inline JS::Scalar::Type Type() const {
     MOZ_ASSERT(this->inited());
     return mType;
   }
 };
 
-typedef TypedArray<int8_t, js::UnwrapInt8Array, JS_GetInt8ArrayData,
-                   js::GetInt8ArrayLengthAndData, JS_NewInt8Array>
-    Int8Array;
-typedef TypedArray<uint8_t, js::UnwrapUint8Array, JS_GetUint8ArrayData,
-                   js::GetUint8ArrayLengthAndData, JS_NewUint8Array>
-    Uint8Array;
-typedef TypedArray<
-    uint8_t, js::UnwrapUint8ClampedArray, JS_GetUint8ClampedArrayData,
-    js::GetUint8ClampedArrayLengthAndData, JS_NewUint8ClampedArray>
-    Uint8ClampedArray;
-typedef TypedArray<int16_t, js::UnwrapInt16Array, JS_GetInt16ArrayData,
-                   js::GetInt16ArrayLengthAndData, JS_NewInt16Array>
-    Int16Array;
-typedef TypedArray<uint16_t, js::UnwrapUint16Array, JS_GetUint16ArrayData,
-                   js::GetUint16ArrayLengthAndData, JS_NewUint16Array>
-    Uint16Array;
-typedef TypedArray<int32_t, js::UnwrapInt32Array, JS_GetInt32ArrayData,
-                   js::GetInt32ArrayLengthAndData, JS_NewInt32Array>
-    Int32Array;
-typedef TypedArray<uint32_t, js::UnwrapUint32Array, JS_GetUint32ArrayData,
-                   js::GetUint32ArrayLengthAndData, JS_NewUint32Array>
-    Uint32Array;
-typedef TypedArray<float, js::UnwrapFloat32Array, JS_GetFloat32ArrayData,
-                   js::GetFloat32ArrayLengthAndData, JS_NewFloat32Array>
-    Float32Array;
-typedef TypedArray<double, js::UnwrapFloat64Array, JS_GetFloat64ArrayData,
-                   js::GetFloat64ArrayLengthAndData, JS_NewFloat64Array>
-    Float64Array;
-typedef ArrayBufferView_base<js::UnwrapArrayBufferView,
-                             js::GetArrayBufferViewLengthAndData,
-                             JS_GetArrayBufferViewType>
-    ArrayBufferView;
-typedef TypedArray<uint8_t, JS::UnwrapArrayBufferMaybeShared,
-                   JS::GetArrayBufferMaybeSharedData,
-                   JS::GetArrayBufferMaybeSharedLengthAndData,
-                   JS::NewArrayBuffer>
-    ArrayBuffer;
+using Int8Array = TypedArray<JS::Int8Array>;
+using Uint8Array = TypedArray<JS::Uint8Array>;
+using Uint8ClampedArray = TypedArray<JS::Uint8ClampedArray>;
+using Int16Array = TypedArray<JS::Int16Array>;
+using Uint16Array = TypedArray<JS::Uint16Array>;
+using Int32Array = TypedArray<JS::Int32Array>;
+using Uint32Array = TypedArray<JS::Uint32Array>;
+using Float32Array = TypedArray<JS::Float32Array>;
+using Float64Array = TypedArray<JS::Float64Array>;
+using ArrayBufferView = ArrayBufferView_base<JS_GetArrayBufferViewType>;
+using ArrayBuffer = TypedArray<JS::ArrayBuffer>;
 
 // A class for converting an nsTArray to a TypedArray
 // Note: A TypedArrayCreator must not outlive the nsTArray it was created from.